### DISCUSSION TITLE
Create Helm chart for easier k8s deployment

### DISCUSSION THREAD

**Discussion by l1ghtman2k** [ScoreTrak/ScoreTrak#229] [General]

**Comment by thisisibrahimd** [comment]
Can I help you out on this? I have some experience with Helm chart and k8s deployment.

**Comment by l1ghtman2k** [comment]
> Can I help you out on this? I have some experience with Helm chart and k8s deployment.

I would appreciate that! Let me know If I could be of any help, or if I could clarify anything!

**Comment by thisisibrahimd** [comment]
Looking more into the repo, I believe there are more steps that need to be taken before a proper/scalable helm/k8s deployment can be made.

Here are my suggestion, let me know what you think
- Separate Server and Client repo into different repos as commits can be tagged and docker images can be built and deploy to GitHub container registry as there is no container registry as of yet. Example:
	- ScoreTrak-Server repo, All go files related to the server.
	- ScoreTrak-Client repo, All the react files in the web folder.
- Creating a helm-charts repo in the ScoreTrak Org as I can't and that would be the best place to store future charts. Here are some examples of helm-charts repos.
	- https://github.com/elastic/helm-charts
	- https://github.com/grafana/helm-charts
	- https://github.com/cockroachdb/helm-charts

**Comment by l1ghtman2k** [comment]
This sounds good, I have just created the https://github.com/ScoreTrak/helm-charts repo,

At the moment, with every merge to the main branch, GitHub actions push the client and server docker images to: https://hub.docker.com/u/l1ghtman, but there is no tagging/versioning, which isn't ideal at all. Let me do a quick research on how I could set this up for this Github.

As to splitting ScoreTrak-Server, and ScoreTrak-Client, I have kind of debated on this for a while. Originally, that is how it was until I decided to merge them into one, because I was following this project layout: https://github.com/golang-standards/project-layout/tree/master/web. Though, it seems like your approach makes more sense, given that we can tag/version/package directly in GitHub

**Comment by thisisibrahimd** [comment]
Thanks for creating the repo. Going to structure for later use.

Ohh awesome. tagging is ideal but it won't be the end of the world. 

That project layout is very nice. However, I believe it assume that your go server will be serving the website but in this case, it is not. Therefore, it seems weird to have one github repo produce two docker images. It is possible to make it work but it won't be pretty. I believe this would also complicate putting an envoy proxy between them, I might be wrong on this one tho.

**Comment by l1ghtman2k** [comment]
I fully agree that a single repo shouldn't be creating two docker images, and now that you mentioned, the go backend does not serve the website, so splitting them make sense. 

For now, I have created the ScoreTrak-Client repo, where I will migrate everything from the web/ folder, and in addition, I will set up GitHub actions to push directly to github packages, instead of my personal registry.
(I will keep the ScoreTrak/ScoreTrak repo for the backend for now, and will rename it to ScoreTrak-Server once we migrate out all of the components.)

If we go this way, it might also make sense to create a separate ScoreTrak/Deployment repo, where we can organize things from Deployment/ folder, what are your thoughts?

**Comment by thisisibrahimd** [comment]
Splitting make sense. Looking forward to seeing both container registry. I can help with that as well, if needed. I will keep lookout for when the ScoreTrak/Client repo is migrated if not done already.

That repo makes sense. I was thinking more ScoreTrak/Docs and store all the knowledge there. We could even add the ScoreTrak/Docs repo into the others as a submodules so users can know that it exists.

Thanks for the invite 👍 .

**Comment by l1ghtman2k** [comment]
I just did the following
- Renamed ScoreTrak-Client to client
- migrated web/ to ScoreTrak/client
- removed web/ from ScoreTrak/ScoreTrak
- Setup github actions to create docker images, and push them into github registry on new release for both client and server
- Created https://github.com/ScoreTrak/Docs

For some reason, GitHub yields a wired behavior when creating packages:
For client:
![image](https://user-images.githubusercontent.com/35672535/120116699-6245f700-c157-11eb-8d62-2aa9c4f7b89f.png)
For server:
![image](https://user-images.githubusercontent.com/35672535/120116762-ad600a00-c157-11eb-924a-c6a94d9e0cb9.png)

Even though, both have packages created, the client doesn't display it correctly

**Reply by thisisibrahimd** [reply]
Nice work.

That does look weird, but works fine. Are you having issue pulling the image with the docker cli?

**Reply by thisisibrahimd** [reply]
```
ibrahimd@earth ~ % docker pull docker.pkg.github.com/scoretrak/client/scoretrak-client:v0.1.1
Error response from daemon: unauthorized: Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured.
```

**Reply by l1ghtman2k** [reply]
Wired... It might be the case that somehow the image is private? I will take a look at it right now

**Reply by thisisibrahimd** [reply]
Yeah I believe it may be private.
https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility

**Reply by thisisibrahimd** [reply]
This might be an intended feature 😢 
https://github.community/t/docker-pull-from-public-github-package-registry-fail-with-no-basic-auth-credentials-error/16358/2

I guess dockerhub is the better solution since it allows for anonymous download

**Reply by thisisibrahimd** [reply]
Apologies for not researching more about github container registry, before suggesting it. I could have sworn they were truly public with no auth.
https://github.community/t/docker-pull-from-public-github-package-registry-fail-with-no-basic-auth-credentials-error/16358/70
I guess you can just get the old workflow  yaml and add tagging feature to that.

**Reply by l1ghtman2k** [reply]
All good! I will look into this a bit more, there seems to be a way to do it using  ghcr.io instead

**Reply by l1ghtman2k** [reply]
Seems like using ghcr.io instead of the default container registry worked fine, could you try pulling on your end(my docker is misbehaving atm.), but I think it should be set for anonymous downloads?

**Reply by thisisibrahimd** [reply]
Tried again and same error. Going try in a clean virtual machine

**Reply by thisisibrahimd** [reply]
Never mind, I was stilling using the old link. ghcr works now, thanks.

**Comment by l1ghtman2k** [comment]
Converted this to a discussion, since it makes more sense this way. Also

**Comment by l1ghtman2k** [comment]
On a side note, I am not sure if ScoreTrak needs this, but for Documentation purposes, we can utilize github wiki:
![image](https://user-images.githubusercontent.com/35672535/120119084-c1aa0400-c163-11eb-9301-8976ad57c3a6.png)

I have also added a slack channel in the description of github org, so you are welcome to join(perhaps this is more useful for real-time communication)

**Reply by thisisibrahimd** [reply]
Github wiki will be very useful. That might be best in the Doc repo.
